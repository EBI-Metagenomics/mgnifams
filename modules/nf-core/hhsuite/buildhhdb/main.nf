process HHSUITE_BUILDHHDB {
    tag "$meta.id"
    // label 'process_high'

    container 'quay.io/microbiome-informatics/hh-suite-db-builder:1.0.0'

    input:
    tuple val(meta), path(a3m)

    output:
    tuple val(meta), path("${meta.id}"), emit: hh_db
    path "versions.yml"                , topic: 'versions'

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    ffindex_build -s ${a3m}_a3m.ff{data,index} ${a3m}
    mpirun -np ${task.cpus} ffindex_apply_mpi ${a3m}_a3m.ff{data,index} -i ${a3m}_hhm.ffindex -d ${a3m}_hhm.ffdata -- hhmake -i stdin -o stdout -v 0
    mpirun -np ${task.cpus} cstranslate_mpi -f -x 0.3 -c 4 -I a3m -i ${a3m}_a3m -o ${a3m}_cs219
    sort -k3 -n -r ${a3m}_cs219.ffindex | cut -f1 > sorting.dat
    ffindex_order sorting.dat ${a3m}_hhm.ff{data,index} ${a3m}_hhm_ordered.ff{data,index}
    mv ${a3m}_hhm_ordered.ffindex ${a3m}_hhm.ffindex
    mv ${a3m}_hhm_ordered.ffdata ${a3m}_hhm.ffdata
    ffindex_order sorting.dat ${a3m}_a3m.ff{data,index} ${a3m}_a3m_ordered.ff{data,index}
    mv ${a3m}_a3m_ordered.ffindex ${a3m}_a3m.ffindex
    mv ${a3m}_a3m_ordered.ffdata ${a3m}_a3m.ffdata

    mkdir -p ${prefix}
    mv ${a3m}_* ${prefix}/

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hh-suite: \$(hhblits -h | grep 'HHblits' | sed -n -e 's/.*\\([0-9]\\+\\.[0-9]\\+\\.[0-9]\\+\\).*/\\1/p')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir -p ${prefix}
    touch ${prefix}/test_a3m_a3m.ffdata
    touch ${prefix}/test_a3m_a3m.ffindex
    touch ${prefix}/test_a3m_cs219.ffdata
    touch ${prefix}/test_a3m_cs219.ffindex
    touch ${prefix}/test_a3m_cs219.log1
    touch ${prefix}/test_a3m_hmm.ffdata
    touch ${prefix}/test_a3m_hmm.ffindex

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hh-suite: \$(hhblits -h | grep 'HHblits' | sed -n -e 's/.*\\([0-9]\\+\\.[0-9]\\+\\.[0-9]\\+\\).*/\\1/p')
    END_VERSIONS
    """
}
