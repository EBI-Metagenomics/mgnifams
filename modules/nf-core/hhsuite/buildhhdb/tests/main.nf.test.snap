{
    "test_a3m_cs219.ffindex": {
        "content": [
            "test_a3m_cs219.ffindex:md5,1893a6d1cf11667bbb65520f5e8d6497"
        ],
        "timestamp": "2024-06-10T13:57:44.959353034"
    },
    "test_a3m_cs219.ffdata": {
        "content": [
            "test_a3m_cs219.ffdata:md5,0a03ab702bc958b26fba34f8d1c5fce9"
        ],
        "timestamp": "2024-06-10T13:57:44.956413336"
    },
    "test_a3m_a3m.ffdata": {
        "content": [
            "test_a3m_a3m.ffdata:md5,1d3e01ef4418ae118cce3e43cd5669aa"
        ],
        "timestamp": "2024-06-10T13:57:44.947206544"
    },
    "a3m - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test_a3m_a3m.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_a3m.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.log1:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,ab0aa7b758ce2f2f171fbf1e14339a52"
                ],
                "hh_db": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test_a3m_a3m.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_a3m.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_cs219.log1:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffdata:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_a3m_hmm.ffindex:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ab0aa7b758ce2f2f171fbf1e14339a52"
                ]
            }
        ],
        "timestamp": "2024-06-10T13:57:49.193842411"
    },
    "test_a3m_cs219.log1": {
        "content": [
            "test_a3m_cs219.log.1:md5,979c3846e8b0e1637d8646ca5581daca"
        ],
        "timestamp": "2024-06-10T13:57:44.962639859"
    },
    "test_a3m_a3m.ffindex": {
        "content": [
            "test_a3m_a3m.ffindex:md5,34f597e4cb2d7d3afa4c655146ff3eb0"
        ],
        "timestamp": "2024-06-10T13:57:44.953720066"
    },
    "test_a3m_hmm.ffindex": {
        "content": [
            "test_a3m_hhm.ffindex:md5,d28d458f9df4962c2e8194d1850c1426"
        ],
        "timestamp": "2024-06-10T13:57:44.967162415"
    }
}